upgrade/v110to120/upgrade.go

```diff
@@ -0,0 +1,342 @@
+			volume.Labels = volumeLabels
+			volume.Spec.RecurringJobs = nil
+		}
+		logrus.Infof("Updating %v volume labels to %v", volume.Name, volume.Labels)
+		updatedVolume, err := lhClient.LonghornV1beta1().Volumes(namespace).Update(&volume)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update %v volume", volume.Name)
+		}
+		volume = *updatedVolume
+	}
+	return nil
+}
+
+func translateVolumeRecurringJobToCRs(namespace string, lhClient *lhclientset.Clientset, sharedMap map[string]*types.RecurringJobSpec) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"translate volume recurringJobs to recurringJob CRs")
+	}()
+
+	log := logrus.WithFields(
+		logrus.Fields{
+			"upgrade":   "translate-volume-recurringJobs-to-recurringJob-CRs",
+			"namespace": namespace,
+		},
+	)
+
+	if len(sharedMap) == 0 {
+		log.Debug("Found 0 recurring job needs to be converted to CR")
+		return nil
+	}
+
+	for _, spec := range sharedMap {
+		log.Infof("Creating %v recurring job CR", spec.Name)
+		job := &longhorn.RecurringJob{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: spec.Name,
+			},
+			Spec: *spec,
+		}
+		log.Debugf("Checking if %v recurring job CR already exists", spec.Name)
+		obj, err := lhClient.LonghornV1beta1().RecurringJobs(namespace).Get(job.Name, metav1.GetOptions{})
+		if err == nil {
+			log.Debugf("Recurring job CR already exists %v", obj)
+			continue
+		}
+		if !apierrors.IsNotFound(err) {
+			log.Debugf("Failed to get recurring job %v", spec.Name)
+		}
+		_, err = lhClient.LonghornV1beta1().RecurringJobs(namespace).Create(job)
+		if err != nil {
+			return errors.Wrapf(err, "failed to create recurring job CR with %v", spec)
+		}
+	}
+	return nil
+}
+
+func cleanupAppliedVolumeCronJobs(namespace string, lhClient *lhclientset.Clientset, kubeClient *clientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"cleanup applied volume cron jobs")
+	}()
+
+	log := logrus.WithFields(
+		logrus.Fields{
+			"upgrade":   "cleanup-applied-volume-cron-jobs",
+			"namespace": namespace,
+		},
+	)
+
+	log.Debugf("Listing all volumes")
+	volumeList, err := lhClient.LonghornV1beta1().Volumes(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Debug("Cannot find volumes")
+			return nil
+		}
+		return errors.Wrap(err, "failed to list volumes")
+	}
+
+	propagation := metav1.DeletePropagationBackground
+	cronJobClient := kubeClient.BatchV1beta1().CronJobs(namespace)
+	for _, v := range volumeList.Items {
+		log.Debugf("Listing all cron jobs for volume %v", v.Name)
+		appliedCronJobROs, err := listVolumeCronJobROs(v.Name, namespace, kubeClient)
+		if err != nil {
+			return errors.Wrapf(err, "failed to list all cron jobs for volume %v", v.Name)
+		}
+		for name := range appliedCronJobROs {
+			log.Infof("Deleting %v cronjob job for %v volume", name, v.Name)
+			err := cronJobClient.Delete(name, &metav1.DeleteOptions{PropagationPolicy: &propagation})
```

Comment:
Do the old cron jobs have any dependencies?

Reply:
Fixed.